lib/rpmsg/rpmsg.c

```diff
@@ -262,4 +262,5 @@ void rpmsg_destroy_ept(struct rpmsg_endpoint *ept)
 	metal_mutex_acquire(&rdev->lock);
 	rpmsg_unregister_endpoint(ept);
 	metal_mutex_release(&rdev->lock);
+	memset(ept, 0, sizeof(*ept));
```

Comment:
can set the endpoint address to ANY.
If RPMsg endpoint is statically defined, how about using ANY address for initialization?

Reply:
> can set the endpoint address to ANY.
> If RPMsg endpoint is statically defined, how about using ANY address for initialization?

Yes, we can require the user set the address to ANY after allocation and before rpmsg_create_ept, but the user may(often?) forget to do it,  it's better to use zero here:
1.global/static variable is zero by compiler/linker automatically
2.It's common practice to memset(or zalloc) the allocated object
so no any special initializition is required anymore.

Reply:
Please don't  clean the endpoint here, better to do it in  rpmsg_init_ept. 
for instance  if you create an ept with specific addresses  you can want to reuse the ept later ( for instance suspend resume usecases).

Reply:
> Please don't clean the endpoint here, better to do it in rpmsg_init_ept.
> for instance if you create an ept with specific addresses you can want to reuse the ept later ( for instance suspend resume usecases).

But, if the user call rpmsg_destroy_ept, should they call rpmsg_create_ept again?
Actually, I want to block this type of error too(call rpmsg_destroy_ept, but not call rpmsg_create_ept).

Reply:
> > Please don't clean the endpoint here, better to do it in rpmsg_init_ept.
> > for instance if you create an ept with specific addresses you can want to reuse the ept later ( for instance suspend resume usecases).
> 
> But, if the user call rpmsg_destroy_ept, should they call rpmsg_create_ept again?
Nothing should prevent this. For instance if application would update the local address or the call backs.

> Actually, I want to block this type of error too(call rpmsg_destroy_ept, but not call rpmsg_create_ept yet).
> And rpmsg_destroy_ept has more side effect than zero out memory:
> 1.Send RPMSG_NS_DESTROY message
> 2.Remove the allocation bit in rpmsg_unregister_endpoint
> Both these actions need recover by call rpmsg_create_ept, then we need pass the address to it anyway.
Why this should happen? except a bug in application...if application call rpmsg_init_ept to create the endpoint this should not occurs. My concern is that this add a restriction in OpenAMP to facilitate application robustness... 

Please tell me if I'm wrong, but here's what you want to solve is the problem when an application calls rpmsg_create_ept without calling rpmsg_init_ept first.
In this case, the ept structure is set to 0 and since the local address is 0, there is no address negotiation.

In this case what about testing the ept->cb != NULL instead?  this should prevent a bad initialization in  rpmsg_destroy_ept and rpmsg_destroy_ept function...
Today this callback is mandatory ( no protection in rpmsg_virtio_rx_callback), but no test on it...

Reply:
> > > Please don't clean the endpoint here, better to do it in rpmsg_init_ept.
> > > for instance if you create an ept with specific addresses you can want to reuse the ept later ( for instance suspend resume usecases).

memory clean only happen in rpmsg_destroy_ept, I can't understand why resume/suspend will call rpmsg_destroy_ept. Even suspend really need call rpmsg_destroy_ept, resume must call rpmsg_create_ept again to recover the side effect done by rpmsg_destroy_ept to make the endpoint work again. As you can see both rpmsg_create_ept and rpmsg_init_ept need the caller pass the local and remote address, the caller should save the address somewhere anyway.

> > 
> > 
> > But, if the user call rpmsg_destroy_ept, should they call rpmsg_create_ept again?
> > Nothing should prevent this. For instance if application would update the local address or the call backs.

But, application shouldn't directly modify the local address since rpmsg_device track the allocation in bitmap field, the unexpected modification make the tracking inaccuracy.

> 
> > Actually, I want to block this type of error too(call rpmsg_destroy_ept, but not call rpmsg_create_ept yet).
> > And rpmsg_destroy_ept has more side effect than zero out memory:
> > 1.Send RPMSG_NS_DESTROY message
> > 2.Remove the allocation bit in rpmsg_unregister_endpoint
> > Both these actions need recover by call rpmsg_create_ept, then we need pass the address to it anyway.
> > Why this should happen? except a bug in application...if application call rpmsg_init_ept to create the endpoint this should not occurs. My concern is that this add a restriction in OpenAMP to facilitate application robustness...
> 
> Please tell me if I'm wrong, but here's what you want to solve is the problem when an application calls rpmsg_create_ept without calling rpmsg_init_ept first.

Since rpmsg_create_ept call rpmsg_init_ept internally, why we need call rpmsg_init_ept then rpmsg_create_ept? is it enough to only call rpmsg_create_ept here?
After grep rpmsg_create_ept and rpmsg_init_ept, you can see all applications just call rpmsg_create_ept, only two places call rpmsg_init_ept:
1.rpmsg_create_ept
2.rpmsg_init_vdev for NS endpoint
So what my patch try to resolve is:
1.rpmsg_endpoint allocate dynamiclly or static
2.before call rpmsg_create_ept
and then application call is_rpmsg_ept_ready, we should return fail here.

> In this case, the ept structure is set to 0 and since the local address is 0, there is no address negotiation.
> 
> In this case what about testing the ept->cb != NULL instead? this should prevent a bad initialization in rpmsg_destroy_ept and rpmsg_destroy_ept function...
> Today this callback is mandatory ( no protection in rpmsg_virtio_rx_callback), but no test on it...

Yes, we can check cb instead, but since is_rpmsg_ept_ready only check the address, I want to make the check more consistent.

Reply:
> > In this case, the ept structure is set to 0 and since the local address is 0, there is no address negotiation.
> > In this case what about testing the ept->cb != NULL instead? this should prevent a bad initialization in rpmsg_destroy_ept and rpmsg_destroy_ept function...
> > Today this callback is mandatory ( no protection in rpmsg_virtio_rx_callback), but no test on it...
> 
> Yes, we can check cb instead, but since is_rpmsg_ept_ready only check the address, I want to make the check more consistent.
No sure that this API is really used... @wjliang: you confirm? 
Have you a use case that justifies the call of this function with a bad ept, after an error returned by the rpmsg_create_ept?

I would like to be sure that we do not introduce restriction in OpenAMP just to help user to debug its application...

Reply:
> > > In this case, the ept structure is set to 0 and since the local address is 0, there is no address negotiation.
> > > In this case what about testing the ept->cb != NULL instead? this should prevent a bad initialization in rpmsg_destroy_ept and rpmsg_destroy_ept function...
> > > Today this callback is mandatory ( no protection in rpmsg_virtio_rx_callback), but no test on it...
> > 
> > 
> > Yes, we can check cb instead, but since is_rpmsg_ept_ready only check the address, I want to make the check more consistent.
> > No sure that this API is really used... @wjliang: you confirm?

My next pull request will use this API, and in our rpmsg driver code also use this API too.

> > Have you a use case that justifies the call of this function with a bad ept, after an error returned by the rpmsg_create_ept?
> 

This patch just fix is_rpmsg_ept_ready return true if it get called before rpmsg_create_ept, since rpmsg_endpoint:
1.normally define as global or static variables
2.or directly embed in a big structure
In both case, rpmsg_endpoint normally initialize to zero, and can't check the validity by compare null pointer like the rpmsg_channel in old version.

> I would like to be sure that we do not introduce restriction in OpenAMP just to help user to debug its application...

The only restriction is that local zero address can't be used anymore, I think it isn't a big issue since:
1.Linux doesn't use zero address too(actually 0-1023 isn't used except 0x35)
2.Zero is too special to use as a valid address just like nobody think zero is a valid pointer
3.The remote side could still use zero address if they really want

Reply:
So is_rpmsg_ept_ready should be kept. 
Now  if you add the "ept->cb != NULL " ( that is missing anyway) in rpmsg_create_ept, this should prevent from a call of rpmsg_create_ept without ept initialization.
In case of error retuurned by rpmsg_create_ept, calling  is_rpmsg_ept_ready is not correct. Adding test on null address seems here over-protection.
On the other size:  if some users has implemented this address as a table index, this patch breaks the legacy
On Linux side:  0 to 1023 address seems not forbidden but reserved for pre-allocated address, means that client could create an endpoint with source address = 0. In this case this could also break the legacy.
For these reasons, I suggest not to add the 0 address restriction but only a check on the CB.
Anyway i'm not the maintainer so let @wjliang decide.

Reply:
I can see the point on @xiaoxiang781216 's purpose, "0" is generic default value for static initialization. And I also agree with @arnop2 to prefer not to add restriction.
This change will only impact the application needs dynamic endpoints allocation, for static endpoints, it can still be 0.  Unless, there are some application strictly expect the endpoint comes from the name service needs to be 0. And thus, I am fine with the idea, cannot see it  breaks. But as this does add restriction, I prefer to get all the reviewers agreement.

But please take @arnop2's suggestion not to set the endpoints address to 0 in destroy, as some user application already assume the previous allocated endpoint address are preserved in the endpoint data struct. You can memset it to 0 after you call destroy.

Reply:
Ok, I will remove zero in the next patch